proof/refine/ARM/Refine.thy

```diff
@@ -322,7 +328,8 @@ lemma valid_sched_init[simp]:
 
 lemma valid_domain_list_init[simp]:
   "valid_domain_list init_A_st"
-  by (simp add: init_A_st_def ext_init_def valid_domain_list_def)
+  apply (simp add: init_A_st_def ext_init_def valid_domain_list_def)
+  sorry
```

Comment:
All of this comes down to a bound on the maximum time the kernel can run, which is determined by the overflow conditions on these constants. Below that bound, everything is well behaved, and deriving monotonicity etc should be easy.

We then "only" need to prove that if the timer never returns something greater than this time bound, everything checks out.

Reply:
To dump some more information here:
- `usToTicks` in C for most platforms is just something of the from `return us * TICKS_IN_US`.
- on Arm, which has slower boards, there is also
```c
#if USE_KHZ
    /* reciprocal division overflows too quickly for dividing by KHZ_IN_MHZ.
     * This operation isn't  used frequently or on many platforms, so use manual
     * division here */
    return div64(us * TIMER_CLOCK_KHZ, KHZ_IN_MHZ);
#else
```
where
```c
#define USE_KHZ (TIMER_CLOCK_HZ % HZ_IN_MHZ > 0)
```
where 
`TIMER_CLOCK_HZ` is set as `TIMER_FREQUENCY` in config, and currently has settings like these:
```
src/plat/allwinnerA20/config.cmake:        TIMER_FREQUENCY 24000000
src/plat/am335x/config.cmake:        TIMER_FREQUENCY 24000000
src/plat/apq8064/config.cmake:        TIMER_FREQUENCY 7000000
src/plat/ariane/config.cmake:        TIMER_FREQUENCY 25000000 PLIC_MAX_NUM_INT 30
src/plat/bcm2711/config.cmake:    # Linux Kernel under clk_osc, thus TIMER_FREQUENCY = 54000000.
src/plat/bcm2711/config.cmake:        TIMER_FREQUENCY 54000000
src/plat/bcm2837/config.cmake:        TIMER_FREQUENCY 19200000
src/plat/exynos4/config.cmake:        TIMER_FREQUENCY 24000000
src/plat/exynos5/config.cmake:        TIMER_FREQUENCY 24000000
src/plat/fvp/config.cmake:        TIMER_FREQUENCY 100000000
src/plat/hifive/config.cmake:        TIMER_FREQUENCY 1000000 PLIC_MAX_NUM_INT 53
src/plat/hikey/config.cmake:        TIMER_FREQUENCY 1200000
src/plat/imx31/config.cmake:        TIMER_FREQUENCY ${TimerFrequency}
src/plat/imx6/config.cmake:        TIMER_FREQUENCY 498000000
src/plat/imx7/config.cmake:        TIMER_FREQUENCY 8000000
src/plat/imx8m-evk/config.cmake:        TIMER_FREQUENCY 8000000
src/plat/odroidc2/config.cmake:        TIMER_FREQUENCY 24000000
src/plat/odroidc4/config.cmake:        TIMER_FREQUENCY 24000000
src/plat/omap3/config.cmake:        TIMER_FREQUENCY 13000000
src/plat/polarfire/config.cmake:        TIMER_FREQUENCY 10000000 PLIC_MAX_NUM_INT 186
src/plat/qemu-arm-virt/config.cmake:        TIMER_FREQUENCY 62500000
src/plat/rocketchip/config.cmake:        TIMER_FREQUENCY 10000000 PLIC_MAX_NUM_INT 0
src/plat/rockpro64/config.cmake:        TIMER_FREQUENCY 24000000
src/plat/spike/config.cmake:        TIMER_FREQUENCY 10000000 PLIC_MAX_NUM_INT 0
src/plat/tk1/config.cmake:        TIMER_FREQUENCY 12000000
src/plat/tx1/config.cmake:        TIMER_FREQUENCY 12000000
src/plat/tx2/config.cmake:        TIMER_FREQUENCY 31250000
src/plat/zynq7000/config.cmake:        TIMER_FREQUENCY 320000000
src/plat/zynqmp/config.cmake:        TIMER_FREQUENCY 100000000
src/plat/allwinnerA20/config.cmake:        TIMER_FREQUENCY 24000000
src/plat/am335x/config.cmake:        TIMER_FREQUENCY 24000000
src/plat/apq8064/config.cmake:        TIMER_FREQUENCY 7000000
src/plat/ariane/config.cmake:        TIMER_FREQUENCY 25000000 PLIC_MAX_NUM_INT 30
src/plat/bcm2711/config.cmake:    # Linux Kernel under clk_osc, thus TIMER_FREQUENCY = 54000000.
src/plat/bcm2711/config.cmake:        TIMER_FREQUENCY 54000000
src/plat/bcm2837/config.cmake:        TIMER_FREQUENCY 19200000
src/plat/exynos4/config.cmake:        TIMER_FREQUENCY 24000000
src/plat/exynos5/config.cmake:        TIMER_FREQUENCY 24000000
src/plat/fvp/config.cmake:        TIMER_FREQUENCY 100000000
src/plat/hifive/config.cmake:        TIMER_FREQUENCY 1000000 PLIC_MAX_NUM_INT 53
src/plat/hikey/config.cmake:        TIMER_FREQUENCY 1200000
src/plat/imx31/config.cmake:        TIMER_FREQUENCY ${TimerFrequency}
src/plat/imx6/config.cmake:        TIMER_FREQUENCY 498000000
src/plat/imx7/config.cmake:        TIMER_FREQUENCY 8000000
src/plat/imx8m-evk/config.cmake:        TIMER_FREQUENCY 8000000
src/plat/odroidc2/config.cmake:        TIMER_FREQUENCY 24000000
src/plat/odroidc4/config.cmake:        TIMER_FREQUENCY 24000000
src/plat/omap3/config.cmake:        TIMER_FREQUENCY 13000000
src/plat/polarfire/config.cmake:        TIMER_FREQUENCY 10000000 PLIC_MAX_NUM_INT 186
src/plat/qemu-arm-virt/config.cmake:        TIMER_FREQUENCY 62500000
src/plat/rocketchip/config.cmake:        TIMER_FREQUENCY 10000000 PLIC_MAX_NUM_INT 0
src/plat/rockpro64/config.cmake:        TIMER_FREQUENCY 24000000
src/plat/spike/config.cmake:        TIMER_FREQUENCY 10000000 PLIC_MAX_NUM_INT 0
src/plat/tk1/config.cmake:        TIMER_FREQUENCY 12000000
src/plat/tx1/config.cmake:        TIMER_FREQUENCY 12000000
src/plat/tx2/config.cmake:        TIMER_FREQUENCY 31250000
src/plat/zynq7000/config.cmake:        TIMER_FREQUENCY 320000000
src/plat/zynqmp/config.cmake:        TIMER_FREQUENCY 100000000
```
(basically anything with a timer frequency that is not a multiple of 1GHz uses the division)

Reply:
So here is an idea:

Instead of all the axioms, we should define `us_to_ticks us = (us * factor1) / factor2`, and then make assumptions about `factor1` and `factor2`. If we are in just a multiplication, we can pick `factor2=1`, and if we have the KHz form we can pick the exact config options.

Reply:
I've had a quick look at this. I have

```
axiomatization us_to_ticks :: "ticks => ticks" where
  conv: "us_to_ticks us = (us * factor1) div factor2"
  and well_defined: "factor2 ?? 0"
  and non_trivial: "0 < 4 * unat (us_to_ticks kernelWCET_us)"

lemma us_to_ticks_mono:
  "mono us_to_ticks"
  using conv well_defined non_trivial
  by blast

lemma us_to_ticks_zero:
  "us_to_ticks 0 = 0"
  apply (subst conv)
  by auto

lemma us_to_ticks_nonzero:
  "y ~= 0 ==> us_to_ticks y ~= 0"
  using well_defined
  by simp

lemma kernelWCET_ticks_no_overflow: 
  "4 * unat (us_to_ticks kernelWCET_us) <= unat max_time"
  using conv well_defined non_trivial
  by blast

lemma us_to_ticks_mult: 
  "unat n * unat (us_to_ticks a) <= unat max_time ==> n * us_to_ticks a = us_to_ticks (n * a)"
  using conv well_defined non_trivial
  by auto

lemma getCurrentTime_buffer_nonzero: "0 < us_to_ticks kernelWCET_us + 5 * (us_to_ticks MAX_PERIOD_US)"
  using conv well_defined non_trivial
  by blast
```

But something seems off. For instance, the proof of `us_to_ticks_mono` doesn't actually use the definition of `mono`, according to `apply_trace`. Are the axioms already contradictory? Should we be able to derive monotonicity already? Can't, for some choice of `factor1` and `factor2` this end up overflowing?

Reply:
If you can derive `us_to_ticks_nonzero` there is definitely something wrong. Just take `factor1 = 1 << 63` and `us = 2`. `us_to_ticks` with that is `0`. Are the old axioms still in scope when you state that lemma?

Oh, I see what is going on: stating it like this means "for all factor1 and factor2", which is immediately false.

Instead of stating `conv` as axiom, try this:
```isabelle
consts' factor1 :: 64 word
consts' factor2 :: 64 word

definition
  "us_to_ticks us = (us * factor1) div factor2"
```
(potentially followed by additional assumptions)

Reply:
I was wondering whether there was a `fixes` or something possible in `axiomatization`, but I couldn't get it working. Is there? If not, how would it be best to add assumptions for this definition? Isn't that exactly what `axiomatization` should be doing for us? I'm guessing we don't want a locale

Reply:
The `consts'`/`definition` paradigm is exactly what we want. `axiomatization` should be avoided as far as humanly possible, because it can introduce inconsistency. `definition` can't.

I don't think we need a locale, `axiomatization` can still talk about `factor1` and `factor2` (should really give those things better names before we actually add them) when they are constants. These should be much simpler-to-check facts than talking about the entire definition of `us_to_ticks`. I.e. saying that `factor2` is not `0` is perfectly fine.